src/utils/roundrobin_poll_adapter.cpp

```diff
@@ -0,0 +1,161 @@
+
+#include "utils/roundrobin_poll_adapter.h"
+
+using std::chrono::milliseconds;
+using std::make_move_iterator;
+
+namespace cppkafka {
+
+RoundRobinPollAdapter::RoundRobinPollAdapter(Consumer& consumer)
+: consumer_(consumer),
+  assignment_callback_(consumer.get_assignment_callback()),
+  revocation_callback_(consumer.get_revocation_callback()),
+  rebalance_error_callback_(consumer.get_rebalance_error_callback()),
+  consumer_queue_(consumer.get_consumer_queue()) {
+    // take over the assignment callback
+    consumer_.set_assignment_callback([this](TopicPartitionList& partitions) {
+        on_assignment(partitions);
+    });
+    // take over the revocation callback
+    consumer_.set_revocation_callback([this](const TopicPartitionList& partitions) {
+        on_revocation(partitions);
+    });
+    // take over the rebalance error callback
+    consumer_.set_rebalance_error_callback([this](Error error) {
+        on_rebalance_error(error);
+    });
+}
+
+RoundRobinPollAdapter::~RoundRobinPollAdapter() {
+    restore_forwarding();
+    //set the original callbacks
+    consumer_.set_assignment_callback(assignment_callback_);
+    consumer_.set_revocation_callback(revocation_callback_);
+    consumer_.set_rebalance_error_callback(rebalance_error_callback_);
+}
+
+void RoundRobinPollAdapter::set_timeout(std::chrono::milliseconds timeout) {
+    consumer_.set_timeout(timeout);
+}
+
+std::chrono::milliseconds RoundRobinPollAdapter::get_timeout() {
+    return consumer_.get_timeout();
+}
+
+Message RoundRobinPollAdapter::poll() {
+    return poll(consumer_.get_timeout());
+}
+
+Message RoundRobinPollAdapter::poll(std::chrono::milliseconds timeout) {
+    bool empty_list = partition_queues_.ref().empty();
+    // Poll group event queue first
+    Message message = consumer_queue_.consume(empty_list ? timeout : milliseconds(0));
+    if (message) {
+        return message;
+    }
+    if (!empty_list) {
+        //consume the next partition
+        message = partition_queues_.next().consume(timeout);
+    }
+    return message;
+}
+
+MessageList RoundRobinPollAdapter::poll_batch(size_t max_batch_size) {
+    return poll_batch(max_batch_size, consumer_.get_timeout());
+}
+
+MessageList RoundRobinPollAdapter::poll_batch(size_t max_batch_size,
+                                              std::chrono::milliseconds timeout) {
+    bool empty_list = partition_queues_.ref().empty();
+    ssize_t remaining_count = max_batch_size;
+    // batch from the group event queue first
+    MessageList messages = consumer_queue_.consume_batch(remaining_count,
+                                                         empty_list ? timeout : milliseconds(0));
+    remaining_count -= messages.size();
+    if ((remaining_count <= 0) || empty_list) {
+        // the entire batch was filled
+        return messages;
+    }
+    // batch from the next partition
+    MessageList partition_messages = partition_queues_.next().consume_batch(remaining_count, timeout);
```

Comment:
I don't timeout twice, the first time checks if the partition number is 0 and in that case only applies the timeout for consumer_queue, otherwise it's 0. 
```
MessageList messages = consumer_queue_.consume_batch(remaining_count,
                                                         empty_list ? timeout : milliseconds(0)); ------>> here
```

Regarding the batching across multiple partitions, I've though about that...not really sure what's best. I felt I should have the same behavior as the single `poll` which only polls one partition per call. It also becomes complicated with the timeout paramter...when you poll multiple partitions under the hood, where do you apply the timeout? On which queue? 
The reason why I added the `get_num_pollable_partitions()` is so that users can poll all partitions until a valid message is found, instead of polling N times with a timeout. In this case you timeout once. 

Example:
```
//current code 
while (1) {
    Message msg = adapter->poll(timeout); //if you have N partitions you get N*timeout potentially
    if (msg) ...                  }

// better alternative (polls all partitions until a msg is found)
while (1) {
    int num = adapter->get_num_pollable_partitions();
    while (num--) {
           Message msg = adapter->poll(); //no timeout
           if (msg) break; 
    }
    std::this_thread::sleep_for(timeout);
}
```

Of course, I could implement this inside `poll` itself, which is fine, but then `poll_batch()` will not work in the same way. Ideally `poll_batch` would poll `max_batch/num_partitions` from each partition so that it's equally distributed. Again, not sure that's best...if you have a preference, let me know.

Reply:
> I don't timeout twice

Fair enough. Does calling poll with timeout 0 work as expected? I imagine so but just checking.

> where do you apply the timeout

You could keep track of how much time it took you to process each queue and decrement that from the timeout. This could actually be applied to the current code, but if it works as it is, that works. Let's keep this simple for now.

Reply:
If no messages ever arrive when you're blocked, won't you stop sending heartbeats (because you're not polling the consumer queue), hence a rebalance will occur?

Reply:
I guess instead of blocking on one of the queues, I could block on the consumer queue, which will ensure that at least heartbeats are sent ?

Reply:
I would just not block. As soon as you hit an empty queue you'll hit the timeout anyway so I don't think there's a need to do that.